File: src/browser/runtimeContinuation.ts
import { buildRuntimeGraph } from "../graph/runtimeGraph.js";
import { computeGlobalRuntimeFingerprint } from "../determinism/globalRuntimeFingerprint.js";
import { computeKaalkaHashPayload } from "../crypto/kaalkaRuntime.js";
import type { ExtractionEnvelope } from "../contracts/runtimeContracts.js";
import { buildBrowserIdentity } from "./browserIdentity.js";
import { captureDom } from "./captureRuntime.js";
import { captureRuntimeSnapshot } from "./runtimeSnapshot.js";
import { restoreRuntimeSession, type RuntimeSessionEnvelope } from "./runtimeSession.js";
import { stabilizeSpaDom } from "./spaStabilizer.js";

export type ContinuationOptions = {
  sessionPath: string;
  encryptionKey: string;
  tick?: number;
};

export async function continueAuthenticatedRuntime(
  url: string,
  options: ContinuationOptions,
): Promise<ExtractionEnvelope> {
  const session = restoreRuntimeSession(options.sessionPath, options.encryptionKey);
  return extractWithSession(url, session, options.tick ?? 0);
}

export async function extractWithSession(
  url: string,
  session: RuntimeSessionEnvelope,
  tick = 0,
): Promise<ExtractionEnvelope> {
  const snapshot = await captureRuntimeSnapshot(url, tick, session);
  const identity = buildBrowserIdentity(snapshot);
  const dom = await captureDom(url);
  const spa = stabilizeSpaDom(dom.html, snapshot.routes[0] ?? "/");

  let playwrightApplied = false;
  try {
    const { chromium } = await import("playwright");
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      extraHTTPHeaders: session.headers ?? {},
    });
    if (Array.isArray(session.cookies) && session.cookies.length > 0) {
      await context.addCookies(session.cookies as Parameters<typeof context.addCookies>[0]);
    }
    const page = await context.newPage();
    if (session.localStorage && Object.keys(session.localStorage).length > 0) {
      await page.addInitScript((storage) => {
        for (const [k, v] of Object.entries(storage as Record<string, string>)) {
          localStorage.setItem(k, v);
        }
      }, session.localStorage);
    }
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30_000 });
    playwrightApplied = true;
    await browser.close();
  } catch {
    playwrightApplied = false;
  }

  const graph = buildRuntimeGraph({
    browser: { url: snapshot.url, dom_hash: snapshot.dom_hash, identity: identity.runtime_identity },
    session: { session_id: session.session_id, continued: playwrightApplied },
    network: snapshot.network.slice(0, 50),
  });

  const envelope: ExtractionEnvelope = {
    bounded: true,
    runtime: {
      available: snapshot.available,
      dom_stabilization: { stabilized_hash: snapshot.dom_hash },
      spa_stabilization: spa,
      session: { session_id: session.session_id, continuation: playwrightApplied },
    },
    browser_ir: {
      runtime_identity: identity.runtime_identity,
      storage: snapshot.storage,
      profile_hash: identity.profile_hash,
    },
    unified_runtime_graph: graph,
    pipeline_hash: computeKaalkaHashPayload({ url, kind: "continuation", tick }),
  };
  envelope.global_runtime_fingerprint = computeGlobalRuntimeFingerprint(envelope, graph);
  return envelope;
}

File: lib/src/browser/runtime_continuation.dart
import '../crypto/kaalka_runtime.dart';
import '../determinism/fingerprint.dart';
import '../graph/runtime_graph.dart';
import 'browser_identity.dart';
import 'capture_runtime.dart';
import 'render_page.dart';
import 'runtime_session.dart';
import 'spa_stabilizer.dart';

/// HTTP-bounded authenticated continuation (Playwright not available in Dart package).
Future<Map<String, dynamic>> continueAuthenticatedRuntime(
  String url, {
  required String sessionPath,
  required String encryptionKey,
  int tick = 0,
}) async {
  final session = restoreRuntimeSession(sessionPath, encryptionKey);
  return extractWithSession(url, session, tick: tick);
}

Future<Map<String, dynamic>> extractWithSession(
  String url,
  Map<String, dynamic> session, {
  int tick = 0,
}) async {
  final rendered = await renderPage(url);
  final html = rendered['html'] as String? ?? '';
  final spa = stabilizeSpaDom(html, route: url);
  final captured = await captureRuntime(url);
  final identity = buildBrowserIdentityFromCapture(captured);
  final graph = buildRuntimeGraph({
    'browser': {'url': url, 'identity': identity['runtime_identity']},
    'session': {'session_id': session['session_id'], 'continued': true},
  });
  final envelope = <String, dynamic>{
    'bounded': true,
    'runtime': {
      'available': rendered['available'] ?? false,
      'dom_stabilization': {'stabilized_hash': captured['dom_hash']},
      'spa_stabilization': spa,
      'session': {'session_id': session['session_id'], 'continuation': true},
    },
    'browser_ir': {
      'runtime_identity': identity['runtime_identity'],
      'storage': captured['storage'],
    },
    'unified_runtime_graph': graph.toJson(),
    'pipeline_hash': computeDeterministicHash(
        {'url': url, 'kind': 'continuation', 'tick': tick}),
  };
  envelope['global_runtime_fingerprint'] =
      computeRuntimePipelineFingerprint(envelope, graph);
  return envelope;
}
